Hash table (OpenAddressingHash)

gravilet opened the pull request:
@DmitriySidnev

gravilet committed:
init OpenAddressingHash

gravilet committed:
code style

gravilet committed:
code style

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+
+class BaseHash {
+ protected:
+    int64_t collection_size;
```
Может быть лучше size_t?

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    int64_t collection_size;
+    static int64_t calculate_hash_function(std::string key,
+                                            int64_t collection_size);
+    explicit BaseHash(int64_t size = 100) : collection_size(size) {}
```
В конструкторе класса лучше использовать не число 100, а константу

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    static int64_t calculate_hash_function(std::string key,
+                                            int64_t collection_size);
+    explicit BaseHash(int64_t size = 100) : collection_size(size) {}
+ public:
```
Должна быть пустая строка перед "public:"

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    // log26(collection_size) - number of significant letters in key
+    // int64_t hash_support;
+    int64_t current_size;
+    int64_t expand_step = collection_size;  // ha-ha-ha!!!
```
Вынести инициализацию переменных в конструктор класса

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    // int64_t hash_support;
+    int64_t current_size;
+    int64_t expand_step = collection_size;  // ha-ha-ha!!!
+ public:
```
Должна быть пустая строка перед "public:"

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    OpenAddressingHash() : BaseHash(100), collection(100), step(1), current_size(0) {}
+    OpenAddressingHash(int64_t size, int64_t step) : BaseHash(100), collection(100),
+                                                     step(step), current_size(0) {}
+    void expand_collection(int64_t new_size);
```
Функция `expand_collection` должна быть private

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+    void insert(std::string key, std::string value);
+    void erase(std::string key);
+    std::string find(std::string key) const;
+    int64_t find_hash(std::string key) const;
```
Сделать private

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+                                                     step(step), current_size(0) {}
+    void expand_collection(int64_t new_size);
+    // int64_t static calculate_hash_function(std::string key);
+    void insert(std::string key, std::string value);
```
Добавить в конце строки override

DmitriySidnev commented on include/hash.h:
```diff
@@ -0,0 +1,67 @@
+//    };
+
+}  // namespace gravilet
+
```
Удалить комментированный код

DmitriySidnev reviewed (changes requested):
Исправить замечания

DmitriySidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,78 @@
+namespace gravilet {
+
+    int64_t BaseHash::calculate_hash_function(std::string key, int64_t collection_size) {
+        int64_t result;
```
Вместо  int64_t должен быть size_t

DmitriySidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,78 @@
+    int64_t BaseHash::calculate_hash_function(std::string key, int64_t collection_size) {
+        int64_t result;
+        // for (char elem : key) {        }
+        result = (key[0] % 26) * collection_size / 26;
```
вместо числа 26 использовать константу

DmitriySidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,78 @@
+
+    std::string OpenAddressingHash::find(std::string key) const {
+        int64_t hash = find_hash(key);
+        return (hash != -1) ? std::get<2>(collection[hash]) : nullptr;
```
разве для типа string возвращаемое значение nullptr допустимо?

DmitriySidnev replied:
Функция `find_hash` гарантирует возвращение значения -1, если `get<0>(collection[hash]) == 0`?

DmitriySidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    void OpenAddressingHash::expand_collection(int64_t new_size) {
+        if (new_size <= this->collection_size)
```
Класс `OpenAddressingHash` - наследник класса `BaseHash` и имеет доступ к полю `collection_size`, this - лишний

DmitriySidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,78 @@
+            std::get<1>(collection_tmp[hash]) = std::get<1>(elem);
+            std::get<2>(collection_tmp[hash]) = std::get<2>(elem);
+        }
+        collection = collection_tmp;
```
Копирование всех элементов из `collection` в `collection_tmp`, а потом обратно. Эффективнее сделать resize collection'а до new_size

gravilet replied:
невозможно при моей реализации - нужно для каждого элемента высчитать его новое положение. при работе в одной `collection` мы можем затереть себе данные на старых позициях

DmitriySidnev commented on test/hash_test.cpp:
```diff
@@ -0,0 +1,11 @@
+
+    EXPECT_EQ("bbb", a.find("20"));
+    EXPECT_EQ(1, a.size());
+}
```
Добавить больше тестов, обязятельно с методом `erase` и ключами со значениями, превышающими размер хэш-таблицы

DmitriySidnev reviewed (changes requested):
There are some comments and mistakes that I found

gravilet committed:
refactoring

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,49 @@
+
+class BaseHash {
+ protected:
+    size_t static const init_hash_size = 100;
```
Я бы назвал default_hash_size.

gravilet replied:
хорошо, заменил

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,49 @@
+    size_t find_hash(std::string key) const;
+
+public:
+    OpenAddressingHash(size_t size = init_hash_size, size_t step = 1) : BaseHash(size),
```
Я думаю так лучше:
```
  OpenAddressingHash(size_t size = init_hash_size, 
                     size_t step = 1) : BaseHash(size), collection(size), step(step), current_size(0) {
  }
```

gravilet replied:
плевать – и так и этак коряво

gravilet commented on include/hash.h:
```diff
@@ -0,0 +1,49 @@
+    size_t find_hash(std::string key) const;
+
+public:
+    OpenAddressingHash(size_t size = init_hash_size, size_t step = 1) : BaseHash(size),
```
плевать – и так и этак коряво

gravilet committed:
refactoring + implement operator[]

gravilet committed:
refactoring operator[]

gravilet committed:
error : ld
undefined reference to `gravilet::OpenAddressingHash<std::string, std::string>::insert(std::string, std::string)'

gravilet committed:
Hash on Templates implemented

gravilet commented:
@zingorn

gravilet committed:
'commit'

alexey-sidnev commented on src/hash.cpp:
```diff
@@ -0,0 +1,100 @@
+
+
+namespace gravilet {
+/*
```
Прибить файл.

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+namespace gravilet {
+
+
+template <class K, class V> class BaseHash {
```
Лучше назвать шаблоны более "длинно".

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+ protected:
+    size_t static const default_hash_size = 100;
+    // for 'string-key' implementation of hash-function based on division on 26
+    size_t static const hash_function_divisor = 26;
```
Зачем?

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+    size_t collection_size;     // size of hash, not a number of elements
+    static size_t calculate_hash_function(K key,
+                                          size_t collection_size) {
+        size_t result;
```
А если я захочу свою функцию хеширования?

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+            std::get<1>(collection_tmp[hash]) = std::get<1>(elem);
+            std::get<2>(collection_tmp[hash]) = std::get<2>(elem);
+        }
+        collection = collection_tmp;
```
`std::move()`

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+        while (std::get<0>(collection[hash])
+               && std::get<1>(collection[hash]) != key
+               && tmp < BaseHash<K, V>::collection_size) {
+            hash += step;
```
```
hash += step % BaseHash<K, V>::collection_size;
```

gravilet replied:
если ты хочешь и от проверки на переполнение в следующих двух строках избавиться, то надо не так. так ты возьмёшь остаток и его прибавишь - и опять возможно переполнение.
все три строки можно заменить на 
`hash = (hash + step) % BaseHash<K, V>::collection_size;`

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+    // find hash - position in collection for the 'key'
+    size_t find_hash(K key) const {
+        size_t hash = BaseHash<K, V>::calculate_hash_function(key, BaseHash<K, V>::collection_size);
+        int tmp = 0;
```
Переименовать

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+            hash = -1;
+        return hash;
+    }
+    V default_value_for_not_found = "";
```
Удалить

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+    size_t size() {
+        return current_size;
+    }
+    int count(K key) const {
```
`isPresent`

alexey-sidnev commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+    V& operator[](K key) {
+        int cnt = count(key);
+        if (cnt == 0) {
+            insert(key, default_value_for_not_found);
```
```
insert(key, V());
```

gravilet commented on include/hash.h:
```diff
@@ -0,0 +1,125 @@
+template <class K, class V> class OpenAddressingHash : BaseHash<K, V> {
+ private:
+    std::vector<std::tuple<bool, K, V>> collection;
+    size_t step;                           // step of Linear Probing
```
сделать проверку в конструкторе, чтоб было взаимно простым с размером контейнера. и, если, плохо, изменять. плюс, пересчитывать при изменении размера контейнера

gravilet reviewed (commented):
про взаимную простоту step и размера

gravilet committed:
coding style
move implementation out of class

gravilet committed:
[unstable] + custom hash-function

gravilet committed:
add lambda-functions
add factory
"const" to "&"
